File: Problem 31.py
memoize = {} # (number, max): possibilities

def coinSums(n, maximum=200):
    if n == 0:
        return 1
    if (n, maximum) not in memoize:
        memoize[n, maximum] = 0
        for coin in [i for i in [1,2,5,10,20,50,100,200] if i <= n and i <= maximum]:
            memoize[n, maximum] += coinSums(n-coin, coin)
    return memoize[n, maximum]
print(coinSums(200))

File: Problem 31.hs
{-# LANGUAGE TupleSections, ScopedTypeVariables, ViewPatterns, FlexibleContexts #-}

import qualified Data.Map.Strict as Map
import Data.Maybe
import Control.Arrow
import Data.STRef
import Control.Monad
import Control.Monad.ST
import Control.Monad.Fix (fix)
import Memoize

coinSums n = runST $ newSTRef Map.empty >>= \ref -> coinSumsM ref (n, 200)

coinSumsM :: STRef s (Map.Map (Int, Int) Int) -> (Int, Int) -> ST s Int
coinSumsM _ (0, _) = return 1
coinSumsM ref (n, max) = fmap sum $ mapM (\coin -> memoizeHelper coinSumsM ref ((n-coin), coin)) $ takeWhile (\i -> i<=n && i<=max) [1,2,5,10,20,50,100,200]

main = print $ coinSums 200